Debug build fails

In my environment, compiling preCICE with CMake in Debug mode fails.
Switching to Release mode leads me to successful compilation.

Here are CMake and Make logs below.
It seems the problem in `assertion` macro.

>   I'm new to both this project and English.
>   In case this issue is against the rule here, please tell me so.
>   When you feel puzzled about what I wrote, please help me improve my language skills.

```
$> cmake -DBUILD_SHARED_LIBS=ON -DPETSC=OFF ..
-- The CXX compiler identification is GNU 7.4.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build configuration: Debug
-- --------------------
-- == CONFIGURATION
-- Library version to build: 1.5.1
-- Build configuration: Debug
-- Build shared libraries: ON
-- Target system: Linux-4.15.0-54-generic
-- Host system: Linux-4.15.0-54-generic
-- Library architecture: x86_64-linux-gnu
-- CXX compiler: /usr/bin/c++
-- CXX compiler flags: <empty>
-- CXX linker: /usr/bin/ld
-- Install prefix: /usr/local
-- Source directory: /path/to/precice
-- Binary directory: /path/to/precice/build
-- --------------------
-- Build with MPI: ON
-- Build with PETSC: OFF
-- Build with PYTHON: ON
-- Configure package generation: ON
-- Install tests/testfiles: ON
-- Additional CTest MPI Flags: <empty>
-- --------------------
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_kill
-- Looking for pthread_kill - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Boost version: 1.65.1
-- Found the following Boost libraries:
--   filesystem
--   log
--   log_setup
--   program_options
--   system
--   thread
--   unit_test_framework
--   date_time
--   regex
--   chrono
--   atomic
-- Found Eigen3: /usr/local/include/eigen3 (Required is at least version "3.2") 
-- Found LibXml2: /usr/lib/x86_64-linux-gnu/libxml2.so (found version "2.9.4") 
-- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- PETSc support disabled
-- Found PythonLibs: /usr/lib/x86_64-linux-gnu/libpython2.7.so (found suitable version "2.7.15+", minimum required is "2.7") 
-- Found PythonInterp: /usr/bin/python2.7 (found version "2.7.15") 
-- Found PythonLibs: /usr/lib/x86_64-linux-gnu/libpython2.7.so (found version "2.7.15+") 
-- Found NumPy: /usr/include (found version "1.13.3") 
-- Compressing changelog
-- Compressing manpage: /path/to/precice/build/packaging/man1/binprecice.1
-- Compressing manpage: /path/to/precice/build/packaging/man1/testprecice.1
-- Adding Test precice.Base
-- Adding Test precice.MPI_Ports
-- Adding Test precice.NoMPI
Debian package generator disabled: Install prefix is not "/usr"
-- Configuring done
-- Generating done
-- Build files have been written to: /path/to/precice/build
```

```
$> make
Scanning dependencies of target precice
[  0%] Building CXX object CMakeFiles/precice.dir/src/action/ComputeCurvatureAction.cpp.o
[  1%] Building CXX object CMakeFiles/precice.dir/src/action/ModifyCoordinatesAction.cpp.o
[  1%] Building CXX object CMakeFiles/precice.dir/src/action/PythonAction.cpp.o
[  2%] Building CXX object CMakeFiles/precice.dir/src/action/ScaleByAreaAction.cpp.o
[  2%] Building CXX object CMakeFiles/precice.dir/src/action/ScaleByDtAction.cpp.o
[  3%] Building CXX object CMakeFiles/precice.dir/src/action/config/ActionConfiguration.cpp.o
[  3%] Building CXX object CMakeFiles/precice.dir/src/com/CommunicateBoundingBox.cpp.o
[  4%] Building CXX object CMakeFiles/precice.dir/src/com/CommunicateMesh.cpp.o
[  4%] Building CXX object CMakeFiles/precice.dir/src/com/Communication.cpp.o
[  5%] Building CXX object CMakeFiles/precice.dir/src/com/ConnectionInfoPublisher.cpp.o
[  5%] Building CXX object CMakeFiles/precice.dir/src/com/MPICommunication.cpp.o
[  6%] Building CXX object CMakeFiles/precice.dir/src/com/MPIDirectCommunication.cpp.o
[  6%] Building CXX object CMakeFiles/precice.dir/src/com/MPIPortsCommunication.cpp.o
[  7%] Building CXX object CMakeFiles/precice.dir/src/com/MPIPortsCommunicationFactory.cpp.o
[  7%] Building CXX object CMakeFiles/precice.dir/src/com/MPIRequest.cpp.o
[  8%] Building CXX object CMakeFiles/precice.dir/src/com/MPISinglePortsCommunication.cpp.o
[  8%] Building CXX object CMakeFiles/precice.dir/src/com/MPISinglePortsCommunicationFactory.cpp.o
[  9%] Building CXX object CMakeFiles/precice.dir/src/com/Request.cpp.o
[  9%] Building CXX object CMakeFiles/precice.dir/src/com/SocketCommunication.cpp.o
[ 10%] Building CXX object CMakeFiles/precice.dir/src/com/SocketCommunicationFactory.cpp.o
[ 10%] Building CXX object CMakeFiles/precice.dir/src/com/SocketRequest.cpp.o
[ 11%] Building CXX object CMakeFiles/precice.dir/src/com/SocketSendQueue.cpp.o
[ 11%] Building CXX object CMakeFiles/precice.dir/src/com/config/CommunicationConfiguration.cpp.o
[ 12%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/BaseCouplingScheme.cpp.o
[ 12%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/CompositionalCouplingScheme.cpp.o
[ 13%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/Constants.cpp.o
[ 13%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/CouplingScheme.cpp.o
[ 14%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/MultiCouplingScheme.cpp.o
[ 14%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/ParallelCouplingScheme.cpp.o
[ 15%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/SerialCouplingScheme.cpp.o
[ 15%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/config/CouplingSchemeConfiguration.cpp.o
[ 16%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/config/PostProcessingConfiguration.cpp.o
[ 16%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/impl/AbsoluteConvergenceMeasure.cpp.o
[ 17%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/impl/AitkenPostProcessing.cpp.o
[ 17%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/impl/BaseQNPostProcessing.cpp.o
[ 18%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/impl/BroydenPostProcessing.cpp.o
[ 18%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/impl/ConstantPreconditioner.cpp.o
[ 19%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/impl/ConstantRelaxationPostProcessing.cpp.o
[ 19%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/impl/HierarchicalAitkenPostProcessing.cpp.o
[ 20%] Building CXX object CMakeFiles/precice.dir/src/cplscheme/impl/IQNILSPostProcessing.cpp.o
In file included from /path/to/precice/src/logging/LogMacros.hpp:12:0,
                 from /path/to/precice/src/logging/Logger.hpp:50,
                 from /path/to/precice/src/utils/Parallel.hpp:5,
                 from /path/to/precice/src/utils/assertion.hpp:12,
                 from /path/to/precice/src/cplscheme/impl/../CouplingData.hpp:4,
                 from /path/to/precice/src/cplscheme/impl/../BaseCouplingScheme.hpp:6,
                 from /path/to/precice/src/cplscheme/impl/PostProcessing.hpp:6,
                 from /path/to/precice/src/cplscheme/impl/BaseQNPostProcessing.hpp:7,
                 from /path/to/precice/src/cplscheme/impl/IQNILSPostProcessing.hpp:3,
                 from /path/to/precice/src/cplscheme/impl/IQNILSPostProcessing.cpp:1:
/path/to/precice/src/utils/prettyprint.hpp: In instantiation of ‘static void pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::printer<U>::print_body(const U&, pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::ostream_type&) [with U = Eigen::Matrix<double, -1, -1>; T = Eigen::Matrix<double, -1, -1>; TChar = char; TCharTraits = std::char_traits<char>; TDelimiters = pretty_print::delimiters<Eigen::Matrix<double, -1, -1>, char>; pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::ostream_type = std::basic_ostream<char>]’:
/path/to/precice/src/utils/prettyprint.hpp:144:35:   required from ‘void pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::operator()(pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::ostream_type&) const [with T = Eigen::Matrix<double, -1, -1>; TChar = char; TCharTraits = std::char_traits<char>; TDelimiters = pretty_print::delimiters<Eigen::Matrix<double, -1, -1>, char>; pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::ostream_type = std::basic_ostream<char>]’
/path/to/precice/src/utils/prettyprint.hpp:217:15:   required from ‘std::basic_ostream<_CharT, _Traits>& pretty_print::operator<<(std::basic_ostream<_CharT, _Traits>&, const pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>&) [with T = Eigen::Matrix<double, -1, -1>; TChar = char; TCharTraits = std::char_traits<char>; TDelimiters = pretty_print::delimiters<Eigen::Matrix<double, -1, -1>, char>]’
/path/to/precice/src/utils/prettyprint.hpp:439:23:   required from ‘typename std::enable_if<pretty_print::is_container<T>::value, std::basic_ostream<_CharT, _Traits>&>::type std::operator<<(std::basic_ostream<_CharT, _Traits>&, const T&) [with T = Eigen::Matrix<double, -1, -1>; TChar = char; TCharTraits = std::char_traits<char>; typename std::enable_if<pretty_print::is_container<T>::value, std::basic_ostream<_CharT, _Traits>&>::type = std::basic_ostream<char>&]’
/path/to/precice/src/cplscheme/impl/IQNILSPostProcessing.cpp:257:7:   required from here
/path/to/precice/src/utils/prettyprint.hpp:117:22: error: ‘void it’ has incomplete type
                 auto it = begin(c);
                      ^~
/path/to/precice/src/utils/prettyprint.hpp:118:28: error: ‘const void the_end’ has incomplete type
                 const auto the_end = end(c);
                            ^~~~~~~
In file included from /usr/local/include/eigen3/Eigen/Core:152:0,
                 from /path/to/precice/src/cplscheme/impl/BaseQNPostProcessing.hpp:3,
                 from /path/to/precice/src/cplscheme/impl/IQNILSPostProcessing.hpp:3,
                 from /path/to/precice/src/cplscheme/impl/IQNILSPostProcessing.cpp:1:
/usr/local/include/eigen3/Eigen/src/Core/StlIterators.h: In instantiation of ‘Eigen::DenseBase<Derived>::const_iterator Eigen::DenseBase<Derived>::cbegin() const [with Derived = Eigen::Matrix<double, -1, -1>; Eigen::DenseBase<Derived>::const_iterator = void]’:
/usr/local/include/eigen3/Eigen/src/Core/StlIterators.h:275:16:   required from ‘Eigen::DenseBase<Derived>::const_iterator Eigen::DenseBase<Derived>::begin() const [with Derived = Eigen::Matrix<double, -1, -1>; Eigen::DenseBase<Derived>::const_iterator = void]’
/usr/include/c++/7/bits/range_access.h:59:27:   required from ‘decltype (__cont.begin()) std::begin(const _Container&) [with _Container = Eigen::Matrix<double, -1, -1>; decltype (__cont.begin()) = void]’
/path/to/precice/src/utils/prettyprint.hpp:117:32:   required from ‘static void pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::printer<U>::print_body(const U&, pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::ostream_type&) [with U = Eigen::Matrix<double, -1, -1>; T = Eigen::Matrix<double, -1, -1>; TChar = char; TCharTraits = std::char_traits<char>; TDelimiters = pretty_print::delimiters<Eigen::Matrix<double, -1, -1>, char>; pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::ostream_type = std::basic_ostream<char>]’
/path/to/precice/src/utils/prettyprint.hpp:144:35:   required from ‘void pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::operator()(pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::ostream_type&) const [with T = Eigen::Matrix<double, -1, -1>; TChar = char; TCharTraits = std::char_traits<char>; TDelimiters = pretty_print::delimiters<Eigen::Matrix<double, -1, -1>, char>; pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>::ostream_type = std::basic_ostream<char>]’
/path/to/precice/src/utils/prettyprint.hpp:217:15:   required from ‘std::basic_ostream<_CharT, _Traits>& pretty_print::operator<<(std::basic_ostream<_CharT, _Traits>&, const pretty_print::print_container_helper<T, TChar, TCharTraits, TDelimiters>&) [with T = Eigen::Matrix<double, -1, -1>; TChar = char; TCharTraits = std::char_traits<char>; TDelimiters = pretty_print::delimiters<Eigen::Matrix<double, -1, -1>, char>]’
/path/to/precice/src/utils/prettyprint.hpp:439:23:   required from ‘typename std::enable_if<pretty_print::is_container<T>::value, std::basic_ostream<_CharT, _Traits>&>::type std::operator<<(std::basic_ostream<_CharT, _Traits>&, const T&) [with T = Eigen::Matrix<double, -1, -1>; TChar = char; TCharTraits = std::char_traits<char>; typename std::enable_if<pretty_print::is_container<T>::value, std::basic_ostream<_CharT, _Traits>&>::type = std::basic_ostream<char>&]’
/path/to/precice/src/cplscheme/impl/IQNILSPostProcessing.cpp:257:7:   required from here
/usr/local/include/eigen3/Eigen/src/Core/StlIterators.h:285:3: error: static assertion failed: YOU_TRIED_CALLING_A_VECTOR_METHOD_ON_A_MATRIX
   EIGEN_STATIC_ASSERT_VECTOR_ONLY(Derived);
   ^
In file included from /usr/local/include/eigen3/Eigen/Core:336:0,
                 from /path/to/precice/src/cplscheme/impl/BaseQNPostProcessing.hpp:3,
                 from /path/to/precice/src/cplscheme/impl/IQNILSPostProcessing.hpp:3,
                 from /path/to/precice/src/cplscheme/impl/IQNILSPostProcessing.cpp:1:
/usr/local/include/eigen3/Eigen/src/Core/StlIterators.h:286:10: error: expression list treated as compound expression in functional cast [-fpermissive]
   return const_iterator(derived(), 0);
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
CMakeFiles/precice.dir/build.make:569: recipe for target 'CMakeFiles/precice.dir/src/cplscheme/impl/IQNILSPostProcessing.cpp.o' failed
make[2]: *** [CMakeFiles/precice.dir/src/cplscheme/impl/IQNILSPostProcessing.cpp.o] Error 1
CMakeFiles/Makefile2:109: recipe for target 'CMakeFiles/precice.dir/all' failed
make[1]: *** [CMakeFiles/precice.dir/all] Error 2
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2
```
<!-- You are very welcome to open issues (mainly for bug reports) here! 
     If you need help building or using preCICE, please contact us:
     https://www.precice.org/resources/#contact
     
     Don't forget to attach your preCICE configuration file!
     
     In order for us to be able to help you more effectively, please also try to include (wherever appropriate):
     * The complete log files or as much as you have available.
       For more informative errors, you may use the logging functionality of preCICE:
       https://github.com/precice/precice/wiki/Logging
     * Any relevant solver configuration
     * Your preCICE version or commit number
     * Information about your platform/OS and dependecies (e.g. MPI implementation and version, PETSc, Boost etc)
-->

--- a/src/mesh/Edge.cpp
+++ b/src/mesh/Edge.cpp
@@ -1,4 +1,5 @@
 #include "Edge.hpp"
+#include "utils/EigenIO.hpp"
 
 namespace precice {
 namespace mesh {
@@ -57,13 +58,12 @@ bool Edge::operator!=(const Edge& other) const
 }
 
 std::ostream& operator<<(std::ostream& stream, const Edge& edge){
-    stream << "LINESTRING (";
-    for (int i = 0; i < 2; i++){
-        stream << edge.vertex(i).getCoords().transpose();
-        if (i < 1)
-            stream << ", ";
-    }
-    return stream << ")";
+    using utils::eigenio::wkt;
+    return stream << "LINESTRING ("
+                  << edge.vertex(0).getCoords().transpose().format(wkt())
+                  << ", "
+                  << edge.vertex(1).getCoords().transpose().format(wkt())
+                  << ')';
 }
 
 }} // namespace precice, mesh

--- a/src/mesh/Quad.cpp
+++ b/src/mesh/Quad.cpp
@@ -1,6 +1,7 @@
 #include "Quad.hpp"
 #include "mesh/Edge.hpp"
 #include "mesh/Vertex.hpp"
+#include "utils/EigenIO.hpp"
 #include <boost/range/concepts.hpp>
 #include <Eigen/Dense>
 #include <Eigen/Geometry>
@@ -168,13 +169,13 @@ bool Quad::operator!=(const Quad& other) const
 
 std::ostream& operator<<(std::ostream& os, const Quad& q)
 {
-    os << "POLYGON ((";
-    for (int i = 0; i < 4; i++){
-        os << q.vertex(i).getCoords().transpose();
-        if (i < 3)
-            os << ", ";
-    }
-    return os <<", " << q.vertex(0).getCoords().transpose() << "))";
+    using utils::eigenio::wkt;
+    return os << "POLYGON (("
+              << q.vertex(0).getCoords().transpose().format(wkt()) << ", "
+              << q.vertex(1).getCoords().transpose().format(wkt()) << ", "
+              << q.vertex(2).getCoords().transpose().format(wkt()) << ", "
+              << q.vertex(3).getCoords().transpose().format(wkt()) << ", "
+              << q.vertex(0).getCoords().transpose().format(wkt()) << "))";
 }
 
 } // namespace mesh

--- a/src/mesh/Triangle.cpp
+++ b/src/mesh/Triangle.cpp
@@ -1,6 +1,7 @@
 #include "Triangle.hpp"
 #include "mesh/Edge.hpp"
 #include "mesh/Vertex.hpp"
+#include "utils/EigenIO.hpp"
 #include <boost/range/concepts.hpp>
 #include <Eigen/Dense>
 #include <Eigen/Geometry>
@@ -123,13 +124,12 @@ bool Triangle::operator!=(const Triangle& other) const
 
 std::ostream& operator<<(std::ostream& os, const Triangle& t)
 {
-    os << "POLYGON ((";
-    for (int i = 0; i < 3; i++){
-        os << t.vertex(i).getCoords().transpose();
-        if (i < 2)
-            os << ", ";
-    }
-    return os <<", " << t.vertex(0).getCoords().transpose() << "))";
+    using utils::eigenio::wkt;
+    return os << "POLYGON (("
+              << t.vertex(0).getCoords().transpose().format(wkt()) << ", "
+              << t.vertex(1).getCoords().transpose().format(wkt()) << ", "
+              << t.vertex(2).getCoords().transpose().format(wkt()) << ", "
+              << t.vertex(0).getCoords().transpose().format(wkt()) << "))";
 }
 
 } // namespace mesh

--- a/src/mesh/Vertex.cpp
+++ b/src/mesh/Vertex.cpp
@@ -1,4 +1,5 @@
 #include "Vertex.hpp"
+#include "utils/EigenIO.hpp"
 
 namespace precice
 {
@@ -47,7 +48,7 @@ void Vertex::tag()
 
 std::ostream &operator<<(std::ostream &os, Vertex const &v)
 {
-  return os << "POINT (" << v.getCoords().transpose() << ")";
+  return os << "POINT (" << v.getCoords().transpose().format(utils::eigenio::wkt()) << ')';
 }
 
 } // namespace mesh

--- a/src/sources.cmake
+++ b/src/sources.cmake
@@ -266,6 +266,7 @@ target_sources(precice
     src/utils/Dimensions.hpp
     src/utils/EigenHelperFunctions.cpp
     src/utils/EigenHelperFunctions.hpp
+    src/utils/EigenIO.hpp
     src/utils/Event.cpp
     src/utils/Event.hpp
     src/utils/EventUtils.cpp
@@ -295,7 +296,6 @@ target_sources(precice
     src/utils/prettyprint.hpp
     src/utils/stacktrace.cpp
     src/utils/stacktrace.hpp
-    src/versions.hpp
     src/versions.hpp
     src/xml/ConfigParser.cpp
     src/xml/ConfigParser.hpp

--- a/src/utils/EigenIO.hpp
+++ b/src/utils/EigenIO.hpp
@@ -1,0 +1,21 @@
+#pragma once
+
+#include <Eigen/Core>
+
+namespace precice {
+namespace utils {
+namespace eigenio {
+
+inline Eigen::IOFormat wkt()
+{
+  return Eigen::IOFormat(
+      Eigen::StreamPrecision,
+      Eigen::DontAlignCols,
+      " ", // Coeff separator
+      ","  // Row separator
+  );
+}
+
+} // namespace eigenio
+} // namespace utils
+} // namespace precice

--- a/src/utils/prettyprint.hpp
+++ b/src/utils/prettyprint.hpp
@@ -38,7 +38,8 @@ namespace pretty_print
         struct has_const_iterator : private sfinae_base
         {
         private:
-            template <typename C> static yes & test(typename C::const_iterator*);
+            template <typename C>
+            static typename std::enable_if<!std::is_void<typename std::iterator_traits<typename C::const_iterator>::value_type>::value, yes &>::type test(typename C::const_iterator*);
             template <typename C> static no  & test(...);
         public:
             static const bool value = sizeof(test<T>(nullptr)) == sizeof(yes);

